Quick_Deploy/HuggingFaceTransformers/README.md

```diff
@@ -0,0 +1,258 @@
+hugging face with Triton.
+
+
+# Next Steps
+The following sections expand on the base tutorial and provide guidance for future sandboxing.
+
+## Loading Cached Models
+In the previous steps, we downloaded the falcon-7b model from hugging face when we
+launched the Triton server. We can avoid this lengthy download process  in subsequent runs
+by loading cached models into Triton. To do so, we can follow the hugging face [tutorial
+for downloading models](https://huggingface.co/docs/hub/models-downloading). Once the model
+is downloaded, we can mount it to the Triton container by adding the following mount option to our
+`docker run` command from earlier (making sure to replace {USER} with your username on
+Linux):
+
+```bash
+# Option to mount a specific cached model (falcon-7b in this case)
+-v /home/{USER}/.cache/huggingface/hub/models--tiiuae--falcon-7b:/root/.cache/huggingface/hub/models--tiiuae--falcon-7b
+
+# Option to mount all cached models on the host system
+-v /home/{USER}/.cache/huggingface:/root/.cache/huggingface
+```
+
+## Triton Tool Ecosystem
+Deploying models in Triton also comes with the benefit of access to a fully-supported suite
+of deployment analyzers to help you better understand and tailor your systems to fit your
+needs. Triton currently has two options for deployment analysis:
+- [Performance Analyzer](https://docs.nvidia.com/deeplearning/triton-inference-server/archives/triton-inference-server-2310/user-guide/docs/user_guide/perf_analyzer.html): An inference performance optimizer.
+- [Model Analyzer](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/model_analyzer.html) A GPU memory and compute utilization optimizer.
+
+### Performance Analyzer
+To use the performance analyzer, please ensure the Triton server is still active with the falcon7b
+model or restart it using the docker run command from above.
+
+Once Triton launches successfully, start a Triton SDK container by running the following in a separate window:
+
+```bash
+docker run -it --net=host nvcr.io/nvidia/tritonserver:23.09-py3-sdk bash
+```
+This container comes with all of Triton's deployment analyzers pre-installed, meaning
+we can simply enter the following to get feedback on our model's inference performance:
+
+```bash
+perf_analyzer -m falcon7b --concurrency-range 1:4
+```
+
+This command should run quickly and profile the performance of our falcon7b model at
+increasing levels of concurrency. As the analyzer runs, it will output useful metrics
+such as latency percentiles, latency by stage of inference, and successful request
+count. Ultimately, the analyzer will neatly summarize the data in the final output:
+
+```json
+Concurrency: 1, throughput: 23.2174 infer/sec, latency 43041 usec
+Concurrency: 2, throughput: 23.3284 infer/sec, latency 85590 usec
+Concurrency: 3, throughput: 23.94 infer/sec, latency 125085 usec
+Concurrency: 4, throughput: 23.773 infer/sec, latency 167879 usec
+```
```

Comment:
Not sure if this is specific feedback, but since the models aren't optimized for batching, the throughput stays flat here and latency gets worse linearly with concurrency - so I don't think we should show concurrency > 1 for now.

Reply:
We were asked to incorporate tangible ways to measure performance and optimize a generic deployment. Seemed like a great time to point to the tools team. But yes, this is a strong point! For now, I've changed it to just collect metrics with concurrency 1. The full detail list the perf analyzer produces is this:
```Request concurrency: 1
  Client:
    Request count: 422
    Throughput: 23.4404 infer/sec
    Avg latency: 42583 usec (standard deviation 3809 usec)
    p50 latency: 41547 usec
    p90 latency: 43611 usec
    p95 latency: 48919 usec
    p99 latency: 61778 usec
    Avg HTTP time: 42576 usec (send/recv 87 usec + response wait 42489 usec)
  Server:
    Inference count: 422
    Execution count: 422
    Successful request count: 422
    Avg request latency: 42140 usec (overhead 24 usec + queue 26 usec + compute input 26 usec + compute infer 41998 usec + compute output 65 usec)

  Server Prometheus Metrics:
    Avg GPU Utilization:
      GPU-57c7b00e-ca04-3876-91e2-c1eae40a0733 : 52.7222%
    Avg GPU Power Usage:
      GPU-57c7b00e-ca04-3876-91e2-c1eae40a0733 : 234.525 watts
    Max GPU Memory Usage:
      GPU-57c7b00e-ca04-3876-91e2-c1eae40a0733 : 35431383040 bytes
    Total GPU Memory:
      GPU-57c7b00e-ca04-3876-91e2-c1eae40a0733 : 51527024640 bytes
Inferences/Second vs. Client Average Batch Latency
Concurrency: 1, throughput: 23.4404 infer/sec, latency 42583 usec```

If you think any of these metrics would be more relevant than the ones I selected, let me know and I can swap them out!

Reply:
Looks great! I haven't actually seen the GPU metric collection feature being used until now, very cool!

Reply:
If you have spare cycles for a couple more runs - I notice the GPU utilization is ~50%. I wonder if we can hit 100% with 2+ instances of the model (if it can fit, I see memory usage is > half).

Reply:
(Disclaimer) This and the previous comment are totally separate from this tutorial PR and only if you have spare cycles/curiosity.

If simply scaling instances doesn't work, another thing to try eventually is batching. Long term we probably want to move away from pipeline to have more fine tuned control over batching, but for a quick example, you could enable dynamic batching on the model, either set high concurrency on perf analyzer or set high max_queue_delay in config.pbtxt to allow larger batch to be gathered, and I think you can do something like this to [batch with the pipeline](https://huggingface.co/docs/transformers/main_classes/pipelines#pipeline-batching):
```python
prompts = []
for request in requsts:
   prompts.append(prompt)

# or some fixed batch size? 
batch_size = len(prompts)
sequences = self.pipeline(prompts, batch_size=batch_size, ...)

responses = []
for seq in sequences:
  # tensor = ...
  # response = ...
  responses.append(response)

return responses
```

Reply:
As a matter of fact, I know it can hit ~100%, because when I was building out the model analyzer part of the tutorial it recommended a config with 2 GPU instances for that very reason. Maybe better to show the output in the MA section to show this very value.